R/Spectra.R

```diff
@@ -1058,14 +1082,61 @@ setReplaceMethod("isolationWindowUpperMz", "Spectra", function(object, value) {
     object
 })
 
+#' @rdname Spectra
+#'
+#' @importFrom methods getFunction
+#'
+#' @exportMethod containsMz
+setMethod("containsMz", "Spectra", function(object, mz = numeric(),
+                                            tolerance = 0,
+                                            ppm = 20, which = c("any", "all"),
+                                            BPPARAM = bpparam()) {
+    which <- match.arg(which)
+    cond_fun <- getFunction(which)
+    if (all(is.na(mz)))
+        return(rep(NA, length(object)))
+    if(is(BPPARAM, "SerialParam"))
+        .has_mz(object, mz, tolerance = tolerance, ppm = ppm,
+                condFun = cond_fun, parallel = BPPARAM)
+    else {
+        sp <- SerialParam()
+        f <- as.factor(dataStorage(object))
+        res <- .lapply(object, FUN = .has_mz, mz = mz, tolerance = tolerance,
+                       condFun = cond_fun, parallel = sp, f = f,
+                       BPPARAM = BPPARAM)
+        unsplit(res, f = f)
+    }
+})
+
+#' @rdname Spectra
+#'
+#' @exportMethod containsNeutralLoss
+setMethod("containsNeutralLoss", "Spectra", function(object, neutralLoss = 0,
+                                                     tolerance = 0, ppm = 20,
+                                                     BPPARAM = bpparam()) {
+    if (is(BPPARAM, "SerialParam")) {
+        .has_mz_each(object, precursorMz(object) - neutralLoss,
+                     tolerance = tolerance, ppm = ppm, parallel = BPPARAM)
+    } else {
+        sp <- SerialParam()
+        f <- as.factor(dataStorage(object))
+        res <- .lapply(object, FUN = function(obj, n, tol, ppm, par) {
+            .has_mz_each(obj, precursorMz(obj) - n, tolerance = tol,
+                         ppm = ppm, parallel = sp)
+        }, n = neutralLoss, tol = tolerance, ppm = ppm, par = sp, f = f,
+                       BPPARAM = BPPARAM)
+        unsplit(res, f = f)
+    }
+})
+
 #' @rdname Spectra
 #'
 #' @exportMethod lapply
-setMethod("lapply", "Spectra", function(X, FUN, ..., BPPARAM = SerialParam()) {
-    if (missing(FUN)) {
+setMethod("lapply", "Spectra", function(X, FUN, f = as.factor(seq_along(X)),
```

Comment:
:+1: I like the idea of exporting `lapply`. Hopefully it is not abused by the users.